### ISSUE TITLE
Two controllers?

### ISSUE THREAD

**Issue by kvaps** [piraeusdatastore/piraeus-operator#52]
Hi, I just tried new piraeus-operator, and found that it is deployed two controllers, one as `kind: Deployment` another one as `kind: StatefulSet`:

```bash
# k get deploy
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
piraeus-op-cs-controller     1/1     1            1           12h
piraeus-op-csi-controller    1/1     1            1           10d
piraeus-op-operator          1/1     1            1           10d
piraeus-op-stork             1/1     1            1           10d
piraeus-op-stork-scheduler   1/1     1            1           10d

# k get sts
NAME                       READY   AGE
piraeus-op-cs-controller   0/1     10d
piraeus-op-etcd            1/1     10d
snapshot-controller        1/1     10d

# k get pod -l app=piraeus-op-cs,role=piraeus-controller
NAME                                        READY   STATUS    RESTARTS   AGE
piraeus-op-cs-controller-0                  1/1     Running   0          5m25s
piraeus-op-cs-controller-75fc99d964-c5z87   1/1     Running   0          12h
```

```bash
root@piraeus-op-cs-controller-0:/# linstor n l
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
┊ Node                                      ┊ NodeType   ┊ Addresses                 ┊ State                     ┊
╞════════════════════════════════════════════════════════════════════════════════════════════════════════════════╡
┊ ks1                                       ┊ SATELLITE  ┊ 172.16.0.6:3366 (PLAIN)   ┊ OFFLINE(OTHER_CONTROLLER) ┊
┊ ks2                                       ┊ SATELLITE  ┊ 172.16.0.7:3366 (PLAIN)   ┊ OFFLINE(OTHER_CONTROLLER) ┊
┊ ks3                                       ┊ SATELLITE  ┊ 172.16.0.8:3366 (PLAIN)   ┊ OFFLINE(OTHER_CONTROLLER) ┊
┊ ks4                                       ┊ SATELLITE  ┊ 172.16.0.9:3366 (PLAIN)   ┊ OFFLINE(OTHER_CONTROLLER) ┊
┊ piraeus-op-cs-controller-0                ┊ CONTROLLER ┊ 10.244.3.216:3366 (PLAIN) ┊ OFFLINE                   ┊
┊ piraeus-op-cs-controller-75fc99d964-c5z87 ┊ CONTROLLER ┊ 10.244.2.152:3366 (PLAIN) ┊ Online                    ┊
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by WanzenBug** [comment]
Hi!

Did you have an existing installation and upgraded it? We recently moved away from using a statefulset for the controller towards using a deployment. My guess would be that the stateful set is a left-over from an older version?

There is a note in the [changelog](https://github.com/piraeusdatastore/piraeus-operator/blob/master/CHANGELOG.md#changed) on this.

**Comment by kvaps** [comment]
Hi, this is new installation, I deployed it from 6b4d020fbe2353e47dd9d7cac7cc3154c0d24e86 using helm chart.
https://github.com/piraeusdatastore/piraeus-operator#deployment-with-helm-v3-chart

The only difference is that maybe I was removing some pods generated by operator afterwards

**Comment by WanzenBug** [comment]
I see. Then it is probably related to a version mismatch. We have the operator image set to `:latest`, however it should instead be `:v0.5.0`. I'm still not sure how it manages to start both a deployment and a statefulset. but maybe this change helps?

The main reason for the change was a node failure on the node with the controller on it. the controller pod itself does not have any internal state, so it is safe to re-schedule it as soon as the other pod is lost.

**Comment by kvaps** [comment]
> The main reason for the change was a node failure on the node with the controller on it. the controller pod itself does not have any internal state, so it is safe to re-schedule it as soon as the other pod is lost.

Please correct me if I'm wrong, but Deployments with `type=Recreate` will never be rescheduled until first pod remove. In this way the are fully identical to StatefulSets.

Have you considered to use [simple leader election](https://kubernetes.io/blog/2016/01/simple-leader-election-with-kubernetes/) and run controller in multiple replicas instead?

**Comment by WanzenBug** [comment]
Actually you are right :) 

And yes, leader election is something we are thinking about

**Comment by kvaps** [comment]
> I see. Then it is probably related to a version mismatch. We have the operator image set to :latest, however it should instead be :v0.5.0. I'm still not sure how it manages to start both a deployment and a statefulset. but maybe this change helps?

I think it explains the everything, first operator was started on one node with old version and generated StatefulSets.
After removal it was recreated on another node in another version, and generated Deployments. Or vise-versa.
In fact I having two controllers running and collision between them.

For now I removed all the StatefulSets except etcd, and restarted controller Deployment.

The most unpleasant thing is that I lost all my data in linstor-controller after this, and I have no idea how it just happened 🙄:

```
root@piraeus-op-cs-controller-75fc99d964-gmrb2:/# linstor n l
╭─────────────────────────────────────────────────────────────────────────────────────────────╮
┊ Node                                      ┊ NodeType   ┊ Addresses                 ┊ State  ┊
╞═════════════════════════════════════════════════════════════════════════════════════════════╡
┊ piraeus-op-cs-controller-75fc99d964-gmrb2 ┊ CONTROLLER ┊ 10.244.1.122:3366 (PLAIN) ┊ Online ┊
╰─────────────────────────────────────────────────────────────────────────────────────────────╯
root@piraeus-op-cs-controller-75fc99d964-gmrb2:/# linstor r l
╭────────────────────────────────────────────────────╮
┊ ResourceName ┊ Node ┊ Port ┊ Usage ┊ Conns ┊ State ┊
╞════════════════════════════════════════════════════╡
╰────────────────────────────────────────────────────╯
```

currently the piraeus-operator pod is running with the from the next docker image:

```
quay.io/piraeusdatastore/piraeus-operator@sha256:20c8211663e71dff46bf53aaf8d726cd116e3bed4f61b8aae432b332238f6e22
```

**Comment by kvaps** [comment]
By the way, I redeployed operator from `:v0.5.0` on piraeus namespace, and I was faced with errors:
```
E0710 14:07:00.939957       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to list *v1.StatefulSet: statefulsets.apps is forbidden: User "system:serviceaccount:piraeus:piraeus-op" cannot list resource "statefulsets" in API group "apps" in the namespace "piraeus"
E0710 14:07:01.945573       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to list *v1.StatefulSet: statefulsets.apps is forbidden: User "system:serviceaccount:piraeus:piraeus-op" cannot list resource "statefulsets" in API group "apps" in the namespace "piraeus"
E0710 14:07:02.950758       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to list *v1.StatefulSet: statefulsets.apps is forbidden: User "system:serviceaccount:piraeus:piraeus-op" cannot list resource "statefulsets" in API group "apps" in the namespace "piraeus"
```

```diff
diff --git a/charts/piraeus/values.yaml b/charts/piraeus/values.yaml
index 581fef8..abdb83f 100644
--- a/charts/piraeus/values.yaml
+++ b/charts/piraeus/values.yaml
@@ -26,7 +26,7 @@ drbdRepoCred: drbdiocred     # <- Specify the kubernetes secret name here
 linstorHttpsControllerSecret: ""  # <- name of secret containing linstor server certificates+key. See docs/security.md
 linstorHttpsClientSecret: ""   # <- name of secret containing linstor client certificates+key. See docs/security.md
 operator:
-  image: "quay.io/piraeusdatastore/piraeus-operator:latest"
+  image: "quay.io/piraeusdatastore/piraeus-operator:v0.5.0"
   controllerSet:
     controllerImage: "quay.io/piraeusdatastore/piraeus-server:v1.7.1"
     luksSecret: ""
```

After fixing rbac:
```diff
diff --git a/charts/piraeus/templates/operator-serviceaccount.yaml b/charts/piraeus/templates/operator-serviceaccount.yaml
index dd749dd..bfad094 100644
--- a/charts/piraeus/templates/operator-serviceaccount.yaml
+++ b/charts/piraeus/templates/operator-serviceaccount.yaml
@@ -32,6 +32,7 @@ rules:
   - apiGroups:
       - apps
     resources:
+      - statefulsets
       - deployments
       - daemonsets
     verbs:
@@ -48,6 +49,7 @@ rules:
       - piraeus-operator
     resources:
       - deployments/finalizers
+      - statefulsets/finalizers
     verbs:
       - update
   - apiGroups:
```

I found that it is generated statefulsets, not deployments:
```bash
# kubectl get sts
NAME                       READY   AGE
piraeus-op-cs-controller   1/1     3m20s
piraeus-op-etcd            1/1     12m
snapshot-controller        1/1     12m
```

**Comment by kvaps** [comment]
Heh, I just found that there were some changes between `master` and `v0.5.0`, rbac problem solved by reinstalling chart from `v0.5.0`

**Comment by WanzenBug** [comment]
Closing this now. The original issue caused by using `:latest` docker images with incompatible versions of the helm chart. This was resolved by pinning the image tag on release.

For actual support of multiple LINSTOR controllers, see #56 and the recently merged #73
